proxy/src/proxy.rs

```diff
@@ -263,12 +263,20 @@ fn handle_server_event(
 
     match event {
         ClientFlowEvent::CommandSent {
-            tag,
             handle: _handle,
+            tag,
         } => {
             // TODO: log handle
             trace!(role = "p2s", ?tag, "---> Forwarded command");
         }
+        ClientFlowEvent::CommandRejected {
+            handle: _handle,
+            tag,
+            status,
+        } => {
+            // TODO: log handle
+            trace!(role = "p2s", ?tag, ?status, "---> Aborted command");
```

Comment:
Do we need to forward the status here? It's complicated, think twice!

Reply:
I think "yes".

But we have an interesting problem here (that I'm sure we discussed at some point but denied it :-))

Situation:

```
// client <--> proxy
C: A1 LOGIN {999...999}
P: + ... // Our proxy never rejects literals on it's own
C: <username> "password"

// proxy <--> server
P: A1 LOGIN {999...999} // forwarded
S: A1 BAD // Server is not happy (but our proxy already accepted!)
```

So, the question is now: How do we make sure to not change the semantics of the connection. And I think: we can't fully. IF we want to have a complete command (in order to have all flexibility to change every aspect of it), we cannot forward "half" commands and need to do the literal dance on each side.

But, luckily, I think, the client should not care too much about *why* a command was rejected. We can also reject it after we received all literals.

Meaning: I think `BAD` should be forwarded as it was before.

Additional:

* I think we should also filter capabilities here so that it is fundamentally not possible to exchange unsupported capabilities in any way with the proxy.

Reply:
Additional:

* I think we should also filter capabilities here so that it is fundamentally not possible to exchange unsupported capabilities in any way with the proxy.

Reply:
Options:

1. Don't send any `Status`:
    - Probably bad idea because it might cause a dead lock in the real client
2. Forward the given `Status::Bad`
    - Is a `::Bad` semantically always okay? Does it contain some text or similar things that might be misleading?
3. Don't send the given `Status::Bad`, but send another `Status` instead
    - Maybe `::No`? Maybe change some meta fields inside the `::Bad`?

Let's wait until Monday for a decision.

Reply:
> 1. Don't send any Status:
> Probably bad idea because it might cause a dead lock in the real client

Yeah :-/ I played around a bit and this is exactly what happens.

> 2. Forward the given Status::Bad
> Is a ::Bad semantically always okay? Does it contain some text or similar things that might be misleading?

Unfortunately, also yeah :-/

When we forward the `Bad` the client could see the following (confusing) trace:

```imap
C: A1 LOGIN {999...999}
P: + ...
C: <username> "password"
S: A1 BAD literal rejected
```

As far as I know, the `text` is the only thing that can make this confusing. But, still, I think this is as close as we can get w/o loosing flexibility.

Worth to mention: From the IMAP RFC:

> A rejected command (BAD response) never changes the state of the connection or of the selected mailbox. A failed command (NO response) generally does not change the state of the connection or of the selected mailbox; the exception being the SELECT and EXAMINE commands.

> 3. Don't send the given Status::Bad, but send another Status instead
> Maybe ::No? Maybe change some meta fields inside the ::Bad?

I'm not sure, really. Hard to overlook the implications. Also given that `No` can potentially change state and `Bad` never does (if we can trust the RFC).

Reply:
Then I'd vote for `::Bad` with a custom text.

Reply:
Oh, good idea :+1:

Reply:
I added complexity. Is it good now?

Reply:
Hehe, I hate that we have to care, but it's beautiful how easy we can make those changes! :-)

I'm not sure if it's a good or bad idea to filter the `code`. My intuition would be to keep changes to a minimum, i.e., remove unsupported capabilities if the `code` is `Code::Capability`, and add a `proxy: please refer to ... ` prefix to the original `text`.

`Alert` is certainly a thing to keep in mind. But are there more codes? I don't know. That's why I feel we should rather forward what we don't know.

Reply:
Resolving as discussed privately.